storage: fatal on replica corruption

tbg opened the pull request:
This path is seldom taken, but when it does and the node keeps running,
it won't be performing as expected. Besides, individual replicas can
become "corrupt" due to problems that affect a larger scope, as seen
in #33375 (which is caused by a txn anomaly, #32784).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/33391)
<!-- Reviewable:end -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33391)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 6706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33391#-LUv8wTu47LidxoE4ZHX:-LUv8wTu47LidxoE4ZHY:b-de31ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e2b1565b8c291790b5e82d19f10ac80e8b7d3e8/pkg/storage/replica.go#L6706)):*
> ```Go
> 		}
> 
> 		log.Fatalf(ctx, "replica is corrupted: %s", pErr)
> ```

Is this the right place to be doing this? Perhaps it would be better to `log.Fatal` in `splitTrigger`. Hmm, looking at where we generate `ReplicaCorruptionError` it seems like that has more to do with replica invariants being violated. I suppose that is corruption, but this isn't what I thought it was (bit corruption on disk). 

Guess I'm fine with doing the `log.Fatal` here.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: fatal on replica corruption

This path is seldom taken, but when it does and the node keeps running,
it won't be performing as expected. Besides, individual replicas can
become "corrupt" due to problems that affect a larger scope, as seen
in #33375 (which is caused by a txn anomaly, #32784).

Release note: None

tbg reviewed (commented):
bors r=petermattis
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33391#-:-LVEQhv11BkyL41xg5YU:bxsp0s9)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 6706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33391#-LUv8wTu47LidxoE4ZHX:-LVEOaDDBf9MuaoN1thN:b-wrksdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e2b1565b8c291790b5e82d19f10ac80e8b7d3e8/pkg/storage/replica.go#L6706)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this the right place to be doing this? Perhaps it would be better to `log.Fatal` in `splitTrigger`. Hmm, looking at where we generate `ReplicaCorruptionError` it seems like that has more to do with replica invariants being violated. I suppose that is corruption, but this isn't what I thought it was (bit corruption on disk). 

Guess I'm fine with doing the `log.Fatal` here.
</blockquote></details>

Yeah, there's more cleanup to do here but the upshots are marginal. I'll leave as-is.


<!-- Sent from Reviewable.io -->
